keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/actuation/ArtifactPromotionListener.kt

```diff
@@ -0,0 +1,56 @@
+package com.netflix.spinnaker.keel.ec2.actuation
+
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.api.ec2.cluster.Cluster
+import com.netflix.spinnaker.keel.api.ec2.image.ArtifactImageProvider
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.clouddriver.model.appVersion
+import com.netflix.spinnaker.keel.events.ResourceDeltaResolved
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.PromotionRepository
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+
+@Component
+@ConditionalOnBean(CloudDriverService::class)
+class ArtifactPromotionListener(
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val promotionRepository: PromotionRepository,
+  private val cloudDriverService: CloudDriverService
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @EventListener(ResourceDeltaResolved::class)
+  fun onDeltaResolved(event: ResourceDeltaResolved) {
```

Comment:
This listener is responsible for marking an artifact version as "has been deployed to environment x"

Reply:
This is cool. Is this only handling "environment has only one cluster" right now?

Reply:
Nope, it should work for multiple clusters. It's listening for when each cluster is resolved so it will mark a version as deployed when the cluster is. There's currently no coordination of "artifact in cluster a needs to be version aligned with artifact in cluster b".

Reply:
Not sure if there's a case for > 1 cluster using the same artifact. I'm not catering for that currently. I guess that _can_ be a thing (e.g. we have special purpose stacks). Need to give that some thought.

Reply:
I think that might be handled by the notion of environments being "stable" (i.e. all resources being in line with desired states). I decided to handle that separately as it needs us to start persisting the current state a little better than the current `resource_event` table. I think this PR is fit for purpose for simpler use-cases without it.

Reply:
Ok sounds good! This all makes sense

Reply:
our cluster model is still regional so 1 cluster per region all using the same artifact (assuming artifact is generically representing the global image (container or imageName + base that we baked)) seems like a legit common state to be considering

Reply:
Raised #361 to track this